lib/facter/custom_facts/core/execution/base.rb

```diff
@@ -100,18 +100,22 @@ def execute_command(command, on_fail, logger = nil, time_limit = nil)
             @log.debug("Executing command: #{command}")
             out, stderr = Open3.popen3(opts, command.to_s) do |_, stdout, stderr, wait_thr|
               pid = wait_thr.pid
-              output = +''
-              err = +''
+              stdout_messages = +''
+              stderr_messages = +''
+              out_reader = Thread.new { stdout.read }
+              err_reader = Thread.new { stderr.read }
               begin
                 Timeout.timeout(time_limit) do
-                  output << stdout.read
-                  err << stderr.read
+                  stdout_messages << out_reader.value
+                  stderr_messages << err_reader.value
```

Comment:
If `stdout.read` (on line 105) raises then ruby will raise that exception when `out_reader.value` is called. So it's possible for line 109 to raise something like `Errno::EIO`, in which case we never call `err_reader.value`. As a result we may never join that thread and fully read `stderr`. I don't know if that's going to be an actual problem though...

Reply:
If `out_reader.value` raises an error, the error message will be logged (if execution is called with `on_fail =>  :raise` line 122). If the stdout will raise an error, the resolvers can't get the information needed for facts to be solved, the stderr message can only be used for debug purposes in this case.